app/lib/frontend/handlers/account.dart

```diff
@@ -57,31 +72,59 @@ Future<shelf.Response> signInCallbackHandler(shelf.Request request) async {
   if (!requestContext.experimentalFlags.useNewSignIn) {
     return notFoundHandler(request);
   }
-  final code = request.requestedUri.queryParameters['code'];
-  if (code != null && code.isNotEmpty) {
-    // TODO: verify state in the response
-    // TODO: verify authuser (=0)
-    // TODO: verify prompt (=none)
-    // TODO: use `expectedNonce` when trying to authenticate `code`.
-    final profile = await authProvider.tryAuthenticateOauthCode(
-      code: code,
-      expectedNonce: '[TBD]',
-    );
-    if (profile == null) {
-      throw AuthenticationException.failed();
-    }
-    // TODO: implement proper action on successful authentication
-    return jsonResponse(
-      {
-        'oauthUserId': '*' * profile.oauthUserId.length,
-        'email': profile.email,
-        'name': profile.name,
-        'imageUrl': profile.imageUrl,
-      },
-      indentJson: true,
+  final params = request.requestedUri.queryParameters;
+  final code = params['code'];
+  if (code == null || code.isEmpty) {
+    return notFoundHandler(request, body: 'Missing `code`.');
+  }
+  if (!requestContext.clientSessionCookieStatus.isPresent) {
+    return notFoundHandler(request, body: 'Missing session cookie.');
+  }
+  final session = await accountBackend.lookupValidUserSession(
+      requestContext.clientSessionCookieStatus.sessionId!);
+  if (session == null) {
+    return notFoundHandler(
+      request,
+      body: 'Expired session.',
+      headers: session_cookie.clearSessionCookies(),
     );
   }
-  return notFoundHandler(request);
+  final expectedNonce = session.openidNonce;
+  if (expectedNonce == null) {
+    return notFoundHandler(request, body: 'Missing `nonce` in session.');
+  }
+  final state = Uri.tryParse('?${params['state'] ?? ''}')?.queryParameters ??
+      <String, String>{};
+  // TODO: verify state in the response
+  // TODO: verify authuser (=0)
+  // TODO: verify prompt (=none)
+  final profile = await authProvider.tryAuthenticateOauthCode(
+    code: code,
+    expectedNonce: expectedNonce,
+  );
+  if (profile == null) {
+    throw AuthenticationException.failed();
+  }
+  await accountBackend.updateClientSessionWithProfile(
+    sessionId: session.sessionId,
+    profile: profile,
+  );
+
+  // TODO: implement proper action on successful authentication
+  final go = state['go'];
+  if (go != null) {
+    // TODO: verify go
+    return redirectResponse(go);
```

Comment:
Maybe, it would be wise to have some sanity checks here?

And not just redirect to anything.

Reply:
nah,